modules/rgbd/test/ocl/test_tsdf.cpp

```diff
@@ -0,0 +1,470 @@
+                        const float Sx = 1.f;   //specular color, can be RGB
+                        const float Lx = 1.f;   //light color
+
+                        Point3f l = normalize(lightPose.translation() - Vec3f(p));
+                        Point3f v = normalize(-Vec3f(p));
+                        Point3f r = normalize(Vec3f(2.f * n * n.dot(l) - l));
+
+                        uchar ix = (uchar)((Ax * Ka * Dx + Lx * Kd * Dx * max(0.f, n.dot(l)) +
+                            Lx * Ks * Sx * specPow<sp>(max(0.f, r.dot(v)))) * 255.f);
+                        color = Vec4b(ix, ix, ix, 0);
+                    }
+
+                    imgRow[x] = color;
+                }
+            }
+        }, nstripes);
+}
+// ----------------------------
+
+static const bool display = false;
+static const bool parallelCheck = false;
+
+class Settings
+{
+public:
+    Ptr<kinfu::Params> params;
+    Ptr<kinfu::Volume> volume;
+    Ptr<Scene> scene;
+    std::vector<Affine3f> poses;
+
+    Settings(bool useHashTSDF, bool onlySemisphere)
+    {
+        if (useHashTSDF)
+            params = kinfu::Params::hashTSDFParams(true);
+        else
+            params = kinfu::Params::coarseParams();
+
+        volume = kinfu::makeVolume(params->volumeType, params->voxelSize, params->volumePose.matrix,
+            params->raycast_step_factor, params->tsdf_trunc_dist, params->tsdf_max_weight,
+            params->truncateThreshold, params->volumeDims);
+
+        scene = Scene::create(params->frameSize, params->intr, params->depthFactor, onlySemisphere);
+        poses = scene->getPoses();
+    }
+};
+
+void displayImage(Mat depth, Mat points, Mat normals, float depthFactor, Vec3f lightPose)
+{
+    Mat image;
+    patchNaNs(points);
+    imshow("depth", depth * (1.f / depthFactor / 4.f));
+    renderPointsNormals(points, normals, image, lightPose);
+    imshow("render", image);
+    waitKey(2000);
+}
+
+void normalsCheck(Mat normals)
+{
+    Vec4f vector;
+    for (auto pvector = normals.begin<Vec4f>(); pvector < normals.end<Vec4f>(); pvector++)
+    {
+        vector = *pvector;
+        if (!cvIsNaN(vector[0]))
+        {
+            float length = vector[0] * vector[0] +
+                vector[1] * vector[1] +
+                vector[2] * vector[2];
+            ASSERT_LT(abs(1 - length), 0.0001f) << "There is normal with length != 1";
+        }
+    }
+}
+
+int counterOfValid(Mat points)
+{
+    Vec4f* v;
+    int i, j;
+    int count = 0;
+    for (i = 0; i < points.rows; ++i)
+    {
+        v = (points.ptr<Vec4f>(i));
+        for (j = 0; j < points.cols; ++j)
+        {
+            if ((v[j])[0] != 0 ||
+                (v[j])[1] != 0 ||
+                (v[j])[2] != 0)
+            {
+                count++;
+            }
+        }
+    }
+    return count;
+}
+
+void normal_test(bool isHashTSDF, bool isRaycast, bool isFetchPointsNormals, bool isFetchNormals)
+{
+    auto normalCheck = [](Vec4f& vector, const int*)
```

Comment:
Please add explicit `setUseOpenCL()` calls here and in `normal_test` function

Reply:
If we have OpenCL we don't need to use `setUseOpenCL`